### PR TITLE
hashing

### PR DESCRIPTION
**Pull request by cgoliver**
Implemented Weisfeiler Lehman graph hashing. (http://jmlr.org/papers/volume12/shervashidze11a/shervashidze11a.pdf)
Given a graph (with/without node and/or edge attributes) produces identical string encodings for isomorphic graphs and has strong guarantees for the converse.

### CODE REVIEW

**Comment by dschult** [comment]
This is good stuff!  Thanks!

You should make a file called tests/test_hashing.py (see the others in that directory for templates). Basically each function with name starting with test will be run by pytest. That replaces the ```if __name__ == "__main__": idiom for testing. 

In naming, abbreviations are long term headaches for us, so we prefer the full name. e.g. the function ```wl_hash``` should probably be: ```weisfeiler_lehman_graph_hash``` and then you could make a shortcut name like:

```python
graph_hash = weisfeiler_lehman_graph_hash
```

In the docs, it might be good to use the term "graph hash" to keep this separated from the hashing needed to store a key in a dict.

Use blank lines in your docstring to indicate structure: one line description (for hyperlinks in html docs) then blank line, then paragraph describing what it does.  Use other nearby files/functions to see the patterns for docs. Combine the description of the example with the example code itself in a section entitled "Examples". References should be included at the bottom and referred to from above. If you have questions ask.. :}

**Comment by cgoliver** [comment]
> This is good stuff! Thanks!
> 
> You should make a file called tests/test_hashing.py (see the others in that directory for templates). Basically each function with name starting with test will be run by pytest. That replaces the ```if **name** == "**main**": idiom for testing.
> 
> In naming, abbreviations are long term headaches for us, so we prefer the full name. e.g. the function `wl_hash` should probably be: `weisfeiler_lehman_graph_hash` and then you could make a shortcut name like:
> 
> ```python
> graph_hash = weisfeiler_lehman_graph_hash
> ```
> 
> In the docs, it might be good to use the term "graph hash" to keep this separated from the hashing needed to store a key in a dict.
> 
> Use blank lines in your docstring to indicate structure: one line description (for hyperlinks in html docs) then blank line, then paragraph describing what it does. Use other nearby files/functions to see the patterns for docs. Combine the description of the example with the example code itself in a section entitled "Examples". References should be included at the bottom and referred to from above. If you have questions ask.. :}

Thanks! Working on it.

Quick question, do you think the module should be named `hashing.py` or `graph_hashing.py`?

**Comment by dschult** [comment]
Let's go with ```graph_hashing```.   Thanks!

**Comment by cgoliver** [comment]
ok should be ready to go. let me know if anything else is missing!

**Comment by dschult** [comment]
I forgot about hooking this into the doc system.... so there's a couple more steps needed.

- in doc/reference/algorithms/index.rst add (alphabetical): graph_hashing
- then create doc/reference/algorithms/graph_hashing.rst to match the others (moral.rst is an example)

You can build the docs locally with "make html" from the doc/ directory. But you should first install stuff in requirements/doc.txt   If you follow the pattern in the rst files you probably don't need to do this.

Also, in the doc_string, put the first doc line on the same line as the """ and leave a blank line after it (to help sphinx identify what is the string to use for short descriptions).

Almost there! :)

**Comment by cgoliver** [comment]
> I forgot about hooking this into the doc system.... so there's a couple more steps needed.
> 
> * in doc/reference/algorithms/index.rst add (alphabetical): graph_hashing
> * then create doc/reference/algorithms/graph_hashing.rst to match the others (moral.rst is an example)
> 
> You can build the docs locally with "make html" from the doc/ directory. But you should first install stuff in requirements/doc.txt If you follow the pattern in the rst files you probably don't need to do this.

`pygraphviz` was giving some issues at install time so I couldn't manage a full doc build, but it seems to have only failed at the rendering stage.

File: networkx/algorithms/graph_hashing.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
+"""
+    Functions for hashing graphs to strings.
+    Isomorphic graphs should be assigned identical hashes.
+    For now, only Weisfeiler-Lehman hashing is implemented.
```
Comment (rossbar) [inline networkx/algorithms/graph_hashing.py:4]:
```suggestion
Functions for hashing graphs to strings.
Isomorphic graphs should be assigned identical hashes.
For now, only Weisfeiler-Lehman hashing is implemented.
```

File: networkx/algorithms/graph_hashing.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    .. [1] Shervashidze, Nino, Pascal Schweitzer, Erik Jan Van Leeuwen,
+       Kurt Mehlhorn, and Karsten M. Borgwardt. Weisfeiler Lehman
+       Graph Kernels. Journal of Machine Learning Research. 2011.
+       http://www.jmlr.org/papers/volume12/shervashidze11a/shervashidze11a.pdf
+    """
+
+    def neighborhood_aggregate(G, node, node_labels, edge_attr=None):
+        """
+            Compute new labels for given node by aggregating
+            the labels of each node's neighbors.
```
Comment (rossbar) [inline networkx/algorithms/graph_hashing.py:103]:
```suggestion
        Compute new labels for given node by aggregating
        the labels of each node's neighbors.
```

File: networkx/algorithms/graph_hashing.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        for nei in G.neighbors(node):
+            prefix = "" if not edge_attr else G[node][nei][edge_attr]
+            label_list.append(prefix + node_labels[nei])
+        return ''.join(sorted(label_list))
+
+    def weisfeiler_lehman_step(G, labels, edge_attr=None, node_attr=None):
+        """
+            Apply neighborhood aggregation to each node
+            in the graph.
+            Computes a dictionary with labels for each node.
```
Comment (rossbar) [inline networkx/algorithms/graph_hashing.py:115]:
```suggestion
        Apply neighborhood aggregation to each node
        in the graph.
        Computes a dictionary with labels for each node.
```

**Comment by dschult** [comment]
I approve this PR...

**Comment by rossbar** [comment]
Fixing up a couple last minor sphinx issues for the docs, then I will merge!

**Comment by rossbar** [comment]
Okay, in this goes - thanks @cgoliver for the cool feature and @dschult for the careful review!

**Comment by cgoliver** [comment]
hey I just noticed the docstring is missing the `nx.` in the function calls:

https://networkx.github.io/documentation/latest/reference/algorithms/generated/networkx.algorithms.graph_hashing.weisfeiler_lehman_graph_hash.html#networkx.algorithms.graph_hashing.weisfeiler_lehman_graph_hash

should I open a new PR?

**Comment by rossbar** [comment]
It's not necessary - I don't see any failures with `pytest --doctest-modules networkx/algorithms`

**Pull request merged**